### PR TITLE
drivers: serial: Add Line Break Detection to stm32

### LINKED ISSUE #41339: stm32, Unable to read UART while checking from Framing error.
I'm implementing DMX512 input on an F4.

Reading it's done at 250000 bauds, the Break and the Mark After Break it's at 100000 so it's common practice to detect a framing error instead of switching baudrate or edge detection. 

I have an interrupt handler for the UART IRQ, inside it I have this piece of code: 

    while (uart_irq_update(dev) && uart_irq_is_pending(dev)) {
       if (uart_irq_rx_ready(dev)) {   
            int err = uart_err_check(dev);
            if (err > 0)
            {
                if(err == UART_ERROR_OVERRUN ){
                    printk("[UART0_RXTX] UART_ERROR_OVERRUN \n");
                }
                if(err == UART_ERROR_PARITY  ){
                    printk("[UART0_RXTX] UART_ERROR_PARITY \n ");
                }
                if(err == UART_ERROR_FRAMING ){
                    printk("[UART0_RXTX] UART_ERROR_FRAMING \n ");
                }
                if(err == UART_BREAK ){
                    printk("[UART0_RXTX] UART_BREAK \n");
                }
                if(err == UART_ERROR_COLLISION ){
                    printk("[UART0_RXTX] UART_ERROR_COLLISION \n");
                }
            }
            int ret = uart_fifo_read(dev,&_rx_buffer,1);
            if (ret > 0)
            {
                printk("read data %d\n"_rx_buffer);
            }
            if (ret < 0)
            {
                printk("error reading data\n");
            }
            else{
                printk("no data\n");   
            }
        }

I should be able to detect the Break or the data but if I detect the Break , uart_fifo_read always returns 0,  this doesn't happen if I try to not detect the break.

I'm expecting to be able to check the error flags and be able to read from the buffer if no error has occured.

### PR DESCRIPTION
**Pull request by andreacampanella**
-Added support for LBD in uart_stm32_err_check with flag clearing.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/41339

Signed-off-by: Andrea Campanella <andrea.campanella@helvar.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@andreacampanella thank you for your contribution!
Please format the commit message as described here:
https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

**Review by erwango — APPROVED** [review]
Approved, but need to fix the commit message to get merged

**Comment by andreacampanella** [comment]
I think the commit message should be formatted as expected now.

**Comment by erwango** [comment]
> I think the commit message should be formatted as expected now.

@andreacampanella Chnage is not visible on our side.
Also, can you fix the trailing white space issue ? 
Cf https://github.com/zephyrproject-rtos/zephyr/runs/4597357992?check_suite_focus=true

**Review by dcpleung — APPROVED** [review]

**Comment by erwango** [comment]
@andreacampanella PR won't be merged unless you fix the compliance issues reported are all fixed.

**Review by FRASTM — APPROVED** [review]

**Comment by andreacampanella** [comment]
Sorry for the delay, I'm honestly very confused by github, I'm used to use different tools and I find in the situation where I don't understand how to change the PR or resubmit it, I can see this changes to the PR comment at the moment: 

-Added support for LBD in uart_stm32_err_check with flag clearing.

Fixes #41339

Signed-off-by: Andrea Campanella andrea.campanella@helvar.com

But it seems that this are not propagated to other users?

**Comment by erwango** [comment]
> Sorry for the delay, I'm honestly very confused by github, I'm used to use different tools and I find in the situation where I don't understand how to change the PR or resubmit it [...]

No problem, this is a frequent situation.
Here is what I propose for this specific case, using "gitk" tool (you'll need to adapt if using git commands or alternate tool, but that should be the same process globally. This being said, I recommend gitk tool):
- "Checkout" the branch you used to submit this PR. From what I can see it should be 'main'. (I suggest to create a specific branch with a specific name each time you want to push a new change, but that's for another time ;-) )
- "Reset branch" to the commit you need to modify (drivers: serial: Add Line Break Detection to stm32)
- Start "git gui"  (File/Start git gui, in gitk). In git gui, click "Amend" the commit 
- Do the required change (in your case, fix the commit message)
- Commit
- Force push using commands (`git push -f ....` )

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -590,6 +590,10 @@ static int uart_stm32_err_check(const struct device *dev)
 		err |= UART_ERROR_FRAMING;
 	}
 
+	if (LL_USART_IsActiveFlag_LBD(UartInstance)) {
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:593]: It seems LDB detection is not available on some F0 variants.
This is then causing build issues on such targets.
To make it work on all SoCs, please put this code under following condition
```
#if !defined(CONFIG_SOC_SERIES_STM32F0X) || defined(USART_LIN_SUPPORT)
```

**Review by erwango — CHANGES_REQUESTED** [review]
One request in order to make code compatible on all series.
Also, please check compliance test failure reported

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -589,11 +589,17 @@ static int uart_stm32_err_check(const struct device *dev)
 	if (LL_USART_IsActiveFlag_FE(UartInstance)) {
 		err |= UART_ERROR_FRAMING;
 	}
-
+	
+	#if !defined(CONFIG_SOC_SERIES_STM32F0X) || defined(USART_LIN_SUPPORT)
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:593]: Thanks for the update.
Please move "#ifdef'ery" to the beginning of the line.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/serial/uart_stm32.c:593]: @andreacampanella Also, please do your change in only one commit as this is one single and coherent change.

**Review by erwango — APPROVED** [review]

**Review by dkalowsk — APPROVED** [review]
Looks good to me, thanks for the contribution!  Should be good for v3.0.0 merging.

**Comment by ycsin** [comment]
@erwango @FRASTM @ABOSTM  Looks like this can be backported?

**Pull request merged**